### ISSUE TITLE
create a streams that deliver more data after an error event

### ISSUE THREAD

**Issue by Peng-Qian** [dart-lang/language#1233]
#### Question
How to create streams that deliver more data after an error event?

According to [dart documentation](https://dart.dev/tutorials/language/streams#error-events), the stream should be able to emit more data after emits an error (add error will not close the stream).
> Streams can also deliver error events like it delivers data events. Most streams will stop after the first error, but it is possible to have streams that deliver more than one error, and streams that deliver more data after an error event.

If a stream needs to be transferred and the transformation may throw an error, is it possible to make the transformed stream not be closed by add error?
```dart
stream.transform(transformer); 

//or 

stream.map(transformation);
```

**Comment by lrhn** [comment]
That's the existing behavior of streams. If you use a stream controller to create the stream, you can keep adding events after adding an error.

Transforming a stream can also preserve both errors and events, as long as the transformer can handle errors.

However, the `async` and `async*` method syntax is not built for handling more than one error. 
An `await for` loop, like `Stream.forEach`, will exit after the first error, and a `yield` cannot emit an error, so most errors from `async*` streams are emitted by throwing and exiting the function, which automatically makes the error the last event of the stream.

For the `yield`, there is a workaround: `yield* Stream.error(error, stackTrace);` will emit an error on the stream without throwing, which allows you to emit more events afterwards.

For `await for`, you do need to encapsulate the error. You can use [`Result.captureStream`](https://pub.dev/documentation/async/latest/async/Result/captureStream.html) to convert a `Stream<T>` to a `Stream<Result<T>>` where the `Result`s will contain both values and errors.

**Comment by Peng-Qian** [comment]
Many thx, @lrhn 

I finally realized the stream is closed by `stream.onErrorReturnWith()` (Rxdart plugin), not `streamController.addError()` or `sink.addError()`.

For `StreamTransformer<S, T> constructor`, the [documentation](https://api.dart.dev/stable/2.9.3/dart-async/StreamTransformer/StreamTransformer.html) says:

> If the transformed stream is listened to multiple times, the onListen callback is called again for each new Stream.listen call. This happens whether the stream is a broadcast stream or not, but the call will usually fail for non-broadcast streams.

For `StreamTransformer.fromHandlers()`, the [documentation](https://api.flutter.dev/flutter/dart-async/StreamTransformer/StreamTransformer.fromHandlers.html) says: 

> Transformers that are constructed this way cannot use captured state if they are used in streams that can be listened to multiple times.

So, basically StreamTransformer should not involve any shared state calculation, and these calculations should be performed by the source/original stream?
